### PR TITLE
[FLINK-17308] Add regular cleanup task for ExecutionGraphCache

### PR DESCRIPTION
**Pull request by tillrohrmann**
## What is the purpose of the change

The WebMonitorEndpoint now schedules are regular cleanup task which
runs every 2 * WebOptions.REFRESH_INTERVAL and tries to clean up
expired ExecutionGraphCache entries. This ensures that we will remove
unused entries.

## Verifying this change

- Added `WebMonitorEndpointTest.cleansUpExpiredExecutionGraphs`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/AbstractExecutionGraphHandler.java [hunk 1]
```diff
@@ -95,7 +96,7 @@ protected AbstractExecutionGraphHandler(
 
 	/**
 	 * Called for each request after the corresponding {@link AccessExecutionGraph} has been retrieved from the
-	 * {@link ExecutionGraphCache}.
+	 * {@link DefaultExecutionGraphCache}.
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/AbstractExecutionGraphHandler.java:99]: should be reverted

File: flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/job/metrics/JobVertexWatermarksHandlerTest.java [hunk 2]
```diff
@@ -87,7 +87,7 @@ public void before() throws Exception {
 			Time.seconds(1),
 			Collections.emptyMap(),
 			metricFetcher,
-			Mockito.mock(ExecutionGraphCache.class),
+			Mockito.mock(DefaultExecutionGraphCache.class),
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/job/metrics/JobVertexWatermarksHandlerTest.java:90]: can we use the noop cache instead?

File: flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/legacy/ExecutionGraphCacheTest.java [hunk 3]
```diff
@@ -59,7 +59,7 @@
 import static org.junit.Assert.fail;
 
 /**
- * Tests for the {@link ExecutionGraphCache}.
+ * Tests for the {@link DefaultExecutionGraphCache}.
  */
 public class ExecutionGraphCacheTest extends TestLogger {
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/legacy/ExecutionGraphCacheTest.java:64]: rename to `DefaultExecutionGraphCacheTest`?

File: flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/WebMonitorEndpoint.java [hunk 4]
```diff
@@ -725,13 +731,28 @@ public WebMonitorEndpoint(
...
 		leaderElectionService.start(this);
+		startExecutionGraphCacheCleanupTask();
+
 		if (hasWebUI) {
 			log.info("Web frontend listening at {}.", getRestBaseUrl());
 		}
 	}
 
+	private void startExecutionGraphCacheCleanupTask() {
+		final long cleanupInterval = 2 * restConfiguration.getRefreshInterval();
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/WebMonitorEndpoint.java:742]: Why did you choose this as the cleanup interval? Why not "just" the refresh interval?

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/AbstractExecutionGraphHandler.java (hunk 1 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/AbstractExecutionGraphHandler.java:99]: Good point. Will revert it.

File: flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/job/metrics/JobVertexWatermarksHandlerTest.java (hunk 2 above)
Comment (tillrohrmann) [reply flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/job/metrics/JobVertexWatermarksHandlerTest.java:90]: I'll check it.

Comment (tillrohrmann) [reply flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/job/metrics/JobVertexWatermarksHandlerTest.java:90]: Yes, it works.

File: flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/legacy/ExecutionGraphCacheTest.java (hunk 3 above)
Comment (tillrohrmann) [reply flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/legacy/ExecutionGraphCacheTest.java:64]: Makes sense. I will change it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/WebMonitorEndpoint.java (hunk 4 above)
Comment (tillrohrmann) [reply flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/WebMonitorEndpoint.java:742]: The idea was that the cleanup does not have to be as aggressive as the refresh interval. Any other multiple of `getRefreshInterval` would have done the same job I guess.

**Comment by tillrohrmann** [comment]
Thanks a lot for the review @zentol. Merging this PR and the backports.

**Pull request closed without merging**